go/database/mpt/forest_test.go

```diff
@@ -2155,6 +2155,147 @@ func TestForest_VisitPathToStorage(t *testing.T) {
 	}
 }
 
+func TestForest_HasEmptyStorage_CastingDoesNotFail(t *testing.T) {
+	for _, variant := range variants {
+		for _, config := range allMptConfigs {
+			for forestConfigName, forestConfig := range forestConfigs {
+				t.Run(fmt.Sprintf("%s-%s-%s", variant.name, config.Name, forestConfigName), func(t *testing.T) {
+					directory := t.TempDir()
+					var err error
+					forest, err := variant.factory(directory, config, forestConfig)
+					if err != nil {
+						t.Fatalf("failed to open forest: %v", err)
+					}
+
+					root := NewNodeReference(ExtensionId(11))
+
+					// Test casting
+
+					isEmpty, err := forest.HasEmptyStorage(&root, common.Address{0x1})
+					if err != nil {
+						t.Fatalf("failed to ask whether account has empty storage: %v", err)
+					}
+					if !isEmpty {
+						t.Error("freshly created account has empty storage, but forest returned true")
+					}
+
+				})
+			}
+		}
+	}
+}
+
+func TestForest_HasEmptyStorage(t *testing.T) {
+	const N = 100
+	for _, variant := range variants {
+		for _, config := range allMptConfigs {
+			for forestConfigName, forestConfig := range forestConfigs {
+				t.Run(fmt.Sprintf("%s-%s-%s", variant.name, config.Name, forestConfigName), func(t *testing.T) {
+					directory := t.TempDir()
+					var err error
+					forest, err := variant.factory(directory, config, forestConfig)
+					if err != nil {
+						t.Fatalf("failed to open forest: %v", err)
+					}
+
+					root := NewNodeReference(EmptyId())
+
+					addresses := getTestAddresses(N)
+					for i, address := range addresses {
+						root, err = forest.SetAccountInfo(&root, address, AccountInfo{Balance: common.Balance{0x1}})
+						if err != nil {
+							t.Fatalf("cannot update account: %v", err)
+						}
+						// fill storage of half of the accounts
+						if i < 50 {
+							continue
+						}
+						root, err = forest.SetValue(&root, address, common.Key{byte(i)}, common.Value{byte(i)})
+						if err != nil {
+							t.Fatalf("cannot update storage: %v", err)
+						}
+					}
+
+					for i := 0; i < N; i++ {
```

Comment:
this loop has a few issues:
* it seems that all conditions starting with `if i >= 60 && i < 70 {` and below are unreachable
* constants 50, 60, 70, ... should be computed from `N`
* I understand that each if condition branch deletes the account by a different means, but why is the call to  `HasEmptyStorage` + consequent assertions repeated every-time? The sequence is always the same (except a different log message)

I think it would be more readable just to divide the list addresses into four sub-slices and for-each on them independently.  
```

for i, address := range addresses[0:N/4] {
 // one way of deleting a value 
}

```
and then the last for-each would check in one iteration that all addresses are empty or not

Reply:
There was a lost `continue` in the code, so the code was unreachable. Thanks for noticing.
Anyhow, I dont think it changes the code that much to compute the constants from `N`. But if you feel like it suits the testing pattern used in the project, I will redo it accordingly.

Reply:
> * I understand that each if condition branch deletes the account by a different means, but why is the call to  `HasEmptyStorage` + consequent assertions repeated every-time? The sequence is always the same (except a different log message)

I think I dont understand this comment. I tried to implement your thought in [this](https://github.com/Fantom-foundation/Carmen/pull/906#discussion_r1636628859) comment.

Reply:
I wanted to say that this block: https://github.com/Fantom-foundation/Carmen/pull/906/files#diff-ebf52a24a2853cb8ec5fd9f7a9be48535f4ff59fc670f0121e51f9c26fec3b89R2190-R2196

is repeated many times.  You iterate over sequences of addresses, but check still the same - i.e. the repeated lines of code.  It would be better to iterate the addresses in one loop and do the test.

Reply:
about the `N` constant, the problem is that it was set to 100 and the conditions rely on this value, but if `N` is changed to 10, all will fall apart.  I.e. it is not much about simplicity, but functionality.